src/openssl.c

```diff
@@ -7811,49 +7826,56 @@ static struct randL_state *randL_getstate(lua_State *L) {
 	return lua_touserdata(L, lua_upvalueindex(1));
 } /* randL_getstate() */
 
-#ifndef HAVE_SYS_SYSCTL_H
-#define HAVE_SYS_SYSCTL_H (BSD || __GLIBC__)
-#endif
-
 #if HAVE_SYS_SYSCTL_H
 #include <sys/sysctl.h> /* CTL_KERN KERN_RANDOM RANDOM_UUID KERN_URND KERN_ARND sysctl(2) */
 #endif
 
 #ifndef HAVE_RANDOM_UUID
-#define HAVE_RANDOM_UUID (HAVE_SYS_SYSCTL_H && defined __linux) /* RANDOM_UUID is an enum, not macro */
-#endif
-
-#ifndef HAVE_KERN_URND
-#define HAVE_KERN_URND (defined KERN_URND)
-#endif
-
-#ifndef HAVE_KERN_ARND
-#define HAVE_KERN_ARND (defined KERN_ARND)
+#define HAVE_RANDOM_UUID (HAVE_SYS_SYSCTL_H && defined __linux__) /* RANDOM_UUID is an enum, not macro */
 #endif
 
 static int randL_stir(struct randL_state *st, unsigned rqstd) {
 	unsigned count = 0;
 	int error;
 	unsigned char data[256];
-#if HAVE_RANDOM_UUID || HAVE_KERN_URND || HAVE_KERN_ARND
-#if HAVE_RANDOM_UUID
+#if defined(HAVE_ARC4RANDOM)
+	while (count < rqstd) {
+		size_t n = MIN(rqstd - count, sizeof data);
+
+		arc4random(data, n);
+
+		RAND_seed(data, n);
+
+		count += n;
+	}
+#elif defined(HAVE_GETRANDOM)
+	while (count < rqstd) {
+ 		size_t n = MIN(rqstd - count, sizeof data);
+
+		n = getrandom(data, n, 0);
+
+		if (n == -1) {
+			break;
+		}
+
+		RAND_seed(data, n);
+
+		count += n;
+	}
+#elif HAVE_RANDOM_UUID
 	int mib[] = { CTL_KERN, KERN_RANDOM, RANDOM_UUID };
-#elif HAVE_KERN_URND
-	int mib[] = { CTL_KERN, KERN_URND };
-#else
-	int mib[] = { CTL_KERN, KERN_ARND };
-#endif
 
 	while (count < rqstd) {
 		size_t n = MIN(rqstd - count, sizeof data);
 
 		if (0 != sysctl(mib, countof(mib), data, &n, (void *)0, 0))
 			break;
 
-		RAND_add(data, n, n);
+		RAND_seed(data, n);
 
 		count += n;
 	}
+
 #endif
```

Comment:
I'm wondering if this `#endif` should actually be an `#else` so that we don't bother with /dev/urandom if we already got randomness from `arc4random`/`getrandom`/`sysctl`